### PR TITLE
promql: fix incorrect results and panics in `sum` and `avg` over mixed custom and exponential buckets, or incompatible custom buckets

### PR DESCRIPTION
**Pull request by charleskorn**
`sum` and `avg` could return incorrect results, or, in the case of `avg`, panic, if mixed custom and exponential buckets were encountered at a single timestamp, or if incompatible custom buckets were encountered at a single timestamp.

### CODE REVIEW

**Comment by krajorama** [comment]
Thank you. Related to https://github.com/prometheus/prometheus/issues/13494

File: promql/promqltest/testdata/native_histograms.test [hunk 1]
```diff
@@ -762,3 +762,36 @@ eval_warn instant at 30s rate(some_metric[30s])
...
+# Test mixing exponential and custom buckets.
+load 6m
+  metric{series="exponential"}         {{sum:4 count:3 buckets:[1 2 1]}}  _                                                                 {{sum:4 count:3 buckets:[1 2 1]}}
+  metric{series="other-exponential"}   {{sum:3 count:2 buckets:[1 1 1]}}  _                                                                 {{sum:3 count:2 buckets:[1 1 1]}}
+  metric{series="custom"}              _                                  {{schema:-53 sum:1 count:1 custom_values:[5 10] buckets:[1]}}     {{schema:-53 sum:1 count:1 custom_values:[5 10] buckets:[1]}}
+  metric{series="other-custom"}        _                                  {{schema:-53 sum:15 count:2 custom_values:[5 10] buckets:[0 2]}}  {{schema:-53 sum:15 count:2 custom_values:[5 10] buckets:[0 2]}}
+
+# T=0: only exponential
+# T=6: only custom
+# T=12: mixed, should be ignored and emit an warning
```
Comment (beorn7) [inline promql/promqltest/testdata/native_histograms.test:775]: an → a

File: promql/engine.go [hunk 2]
```diff
@@ -2784,6 +2784,7 @@ type groupedAggregation struct {
 	seen              bool // Was this output groups seen in the input at this timestamp.
 	hasFloat          bool // Has at least 1 float64 sample aggregated.
 	hasHistogram      bool // Has at least 1 histogram sample aggregated.
+	abandonHistogram  bool // If true, group has seen mixed exponential and custom buckets, or incompatible custom buckets.
```
Comment (beorn7) [inline promql/engine.go:2787]: Maybe let's call it "incompatibleHistograms"? The current name sounds like we'll kick out the histogram, and then we'll still get a valid result or something.

File: promql/promqltest/testdata/native_histograms.test [hunk 3]
```diff
@@ -762,3 +762,36 @@ eval_warn instant at 30s rate(some_metric[30s])
...
+  metric{series="3"} {{schema:-53 sum:1 count:1 custom_values:[5 10] buckets:[1]}} {{schema:-53 sum:1 count:1 custom_values:[5 10] buckets:[1]}} {{schema:-53 sum:1 count:1 custom_values:[5 10] buckets:[1]}}
+
+# T=0: incompatible, should be ignored and emit a warning
+# T=6: compatible
+# T=12: incompatible followed by compatible, should be ignored and emit a warning
+eval_warn range from 0 to 12m step 6m sum(metric)
+  {} _ {{schema:-53 sum:2 count:2 custom_values:[5 10] buckets:[2]}} _
+
+eval_warn range from 0 to 12m step 6m avg(metric)
+  {} _ {{schema:-53 sum:1 count:1 custom_values:[5 10] buckets:[1]}} _
```
Comment (beorn7) [inline promql/promqltest/testdata/native_histograms.test:797]: Could we test the case where just some aggregation results are impossible? Something like `sum by (foo) (metric)`, where for some values of `foo` , I get valid combinations (only floats, or only compatible histograms) and for others I don't (float + histogram, incompatible histograms). So I should get some returned elements _plus_ the warning.

**Review by beorn7 — CHANGES_REQUESTED** [review]
Thank you very much. Looks good, just ideas about naming and more tests.

File: promql/promqltest/testdata/native_histograms.test (hunk 1 above)
Comment (charleskorn) [reply promql/promqltest/testdata/native_histograms.test]: Good catch, fixed in 82bb35f

File: promql/engine.go (hunk 2 above)
Comment (charleskorn) [reply promql/engine.go]: Changed in 82bb35f

File: promql/promqltest/testdata/native_histograms.test (hunk 3 above)
Comment (charleskorn) [reply promql/promqltest/testdata/native_histograms.test:797]: Good idea, added in 5cfdde3

**Review by zenador — APPROVED** [review]
Good catch, thank you!

**Review by beorn7 — DISMISSED** [review]
Thank you very much.

**Comment by beorn7** [comment]
@charleskorn could you resolve the merge conflicts once more? I'll try to merge ASAP after that.

**Comment by charleskorn** [comment]
> @charleskorn could you resolve the merge conflicts once more? I'll try to merge ASAP after that.

Yep, done.

**Review by beorn7 — APPROVED** [review]

**Pull request merged**